### ISSUE TITLE
Bug in Calculating the amount of heat penalty when weapons of different heat multipliers are used

### ISSUE THREAD

**Issue by NavidA1** [lisongmechlab/lsml#764]
Hi

In MWO, when a group of different weapons that are in the same penalty group are fired together, the penalty system considers the highest heat in the group and the highest multiplier in the group to calculate penalty.

In Li Song mechlab currently, the multiplier is chosen from the weapon with the highest heat rather than the highest multiplier present in the group.

Example:
Firing 1xcERPPC + 2xcGauss on a mech with no quirks or skill:
cERPPC: 14.5 heat and 7 penalty multiplier
cGauss: 1 heat and 8.5 penalty multiplier

Firing that weapon combo will net you one stage of penalty at 3 weapons (0.18 multiplier)

In MWO:
Total heat = 14.5 + 2x1 + (0.18 * 8.5 * 14.5) = 38.685

In Li Song:
Total heat = 14.5 + 2x1 + (0.18 * 7 * 14.5) = 34.77

**Comment by LiSongMWO** [comment]
Hello!

The calulations that LSML uses for ghost heat are based on the official post explaining the mechanic from PGI devs here: https://mwomercs.com/forums/topic/127904-heat-scale-the-maths/. 

The tests here verify that LSML computes the same results as the examples given:
[src/test/java/org/lisoft/lsml/model/metrics/GhostHeatTest.java](https://github.com/lisongmechlab/lsml/blob/23792ad6a451a4eed0a396d97dd416e723e18868/src/test/java/org/lisoft/lsml/model/metrics/GhostHeatTest.java)

Can you please give me a link to a source (article, patch notes, somebody doing science, etc...) where I can find more up-to-date info?

**Comment by NavidA1** [comment]
The official post is missing a piece of information.

At the time, the weapons that were put in the same penalty group had the same penalty multiplier.
The case that post has not covered is how penalty is handled when two weapons are in the same penalty group, with one having the higher heat, and the other having the higher multiplier.

The official post states that the highest heat value of the group is chosen for penalty calculation (assuming all the multipliers of the weapons in the group are the same). This usually implies that the multiplier used for penalty calculation is also coming from the weapon that has the highest heat.

However, this is not the case. The penalty algorithm searches for max values two times. Once it searches for higher base heat in the group, and then it searches for the highest penalty multiplier in the group. Then uses those two values to calculate penalty.

In Li Song mechlab, the highest heat weapon is picked (according to the official post), but the heat multiplier is not picking the highest value in the group, but picking it off the weapon that had the highest heat (which is a piece of info missing on the official post).

This only leads to different results when you mix weapons that no only have different heat, but different heat multiplier too. For example Gauss and PPC weapons... where Gauss has the highest multiplier (8.5 gauss vs 7 ppcs)... but PPCs have the highest heat.

No material is available on this as far as I know. It is directly based on tests. A portion of the MWO community know this based on knowledge pass on to them in the past.

**Comment by NavidA1** [comment]
This ties into another issue as well... which is posted here:
https://github.com/lisongmechlab/lsml/issues/767

**Comment by LiSongMWO** [comment]
I'm willing to change this but I need to have better reference/data to rely on than: "It's known among some of the community and passed down from generation to generation". This kind of tribal knowledge is anectdotally unreliable in my experience, it tends to get changed and misinterpreted at every step, and I feel uneasy basing anything on such.

If we can do in game tests that prove that the current algorithm is wrong, and that your proposed one is not incorrect, than that will be sufficient. To do this we need to find a build that has a group of weapons with the same heat group but different multipliers that would cause an error in the computation but where the alpha isn't high enough to overheat. We need to drop into a map with a neutral heat offset and compare the heat% of an alpha with the computed %, taking into account the ramp up animation of the heat bar in respect to the heat sinking. We'd expect the alpha heat to be closer to the proposed caluclation than the old to prove the hypothesis. And before we can science this for real, #767 needs to be resolved as that would impact any tests that can be done.

**Comment by NavidA1** [comment]
I've made a number of tests for this case.

First case is made to confirm the correct value of a mech's heat capacity.
To test how much heat capacity a mech has, I created a case study with 10 double heatsinks (which should provide us with 50 heat pool). The test mech is a Madcat MKII-2 with the following loadout:
https://mwo.nav-alpha.com/mechlab?b=7957d2ba_MCII-2
- 1xLB20
- 1xLB10
- 2xERPPC
- 4xSRM6
- XL350 engine (10 total double heatsinks)

The build above generates exactly a 50 heat spike (no duration based weapons) on an alpha strike. And in theory should not power down the mech since it does not cross 50. This method of test is chosen since the flash-based UI has a few frames of delay before updating to the current heat value, while overheat logic is per frame, untied to the UI, and instant.

Mech having 50 heat capacity is confirmed by the following recording:
https://youtu.be/X7cPqzhKNk8
Any additional heat on top of what is recorded leads to an overheat shutdown.

=========================

with confirmed 50 base heat capacity on a mech as a reference, another test can be designed to verify the the correct amount of generated heat penalty. To do this, a Vapor eagle-1 with the following build is considered:
https://mwo.nav-alpha.com/mechlab?b=ded8f945_VGL-1
The above build has:
- 1xGauss (1 heat and 8.5 multiplier)
- 2xERPPC (each with 14.5 heat and 7 multiplier)
- SRM2 (1.6 heat)
- XL255 (10 total double heatsinks)

The total heat generated by that build after an alpha strike is:
Total heat (MWO) = 14.5 + 14.5 + 1 + (0.18 * 14.5 * 8.5) + 1.6 = 53.785
Which should overheat the mech

According to Li Song, the total heat generated by this build after an alpha strike is:
Total heat (Li Song) = 14.5 + 14.5 + 1 + (0.18 * 14.5 * 7) + 1.6 = 49.87
which should not create an overheat.

The following recording shows that an alpha strike is crossing the 50 heat threshold causing an overheat, indicating that the total heat is higher than 50.
https://youtu.be/hMI40xW5yQg


To confirm the value calculated with the correct multiplier (53.785 total alpha heat), the same Vapor eagle used above is equipped with 8 additional double heatsinks to bring up the total amount of heat capacity to 54, reaching just above the alpha strike heat. In this case an alpha strike should not cause an overheat.
VGL-1 with added heatsinks: https://mwo.nav-alpha.com/mechlab?b=f70662ff_VGL-1

The following recording shows that the VGL with the additional heatsinks is able to alpha strike without an overheat:
https://youtu.be/7T3OeZ2z4M0

**Comment by NavidA1** [comment]
I've made a number of tests for this case (including issue https://github.com/lisongmechlab/lsml/issues/767)

First case is made to confirm the correct value of a mech's heat capacity.
To test how much heat capacity a mech has, I created a case study with 10 double heatsinks (which should provide us with 50 heat pool). The test mech is a Madcat MKII-2 with the following loadout:
https://mwo.nav-alpha.com/mechlab?b=7957d2ba_MCII-2
- 1xLB20
- 1xLB10
- 2xERPPC
- 4xSRM6
- XL350 engine (10 total double heatsinks)

The build above generates exactly a 50 heat spike (no duration based weapons) on an alpha strike. And in theory should not power down the mech since it does not cross 50. This method of test is chosen since the flash-based UI has a few frames of delay before updating to the current heat value, while overheat logic is per frame, untied to the UI, and instant.

Mech having 50 heat capacity is confirmed by the following recording:
https://youtu.be/X7cPqzhKNk8
Any additional heat on top of what is recorded leads to an overheat shutdown.

=========================

with confirmed 50 base heat capacity on a mech as a reference, another test can be designed to verify the the correct amount of generated heat penalty. To do this, a Vapor eagle-1 with the following build is considered:
https://mwo.nav-alpha.com/mechlab?b=ded8f945_VGL-1
The above build has:
- 1xGauss (1 heat and 8.5 multiplier)
- 2xERPPC (each with 14.5 heat and 7 multiplier)
- SRM2 (1.6 heat)
- XL255 (10 total double heatsinks)

The total heat generated by that build after an alpha strike is:
Total heat (MWO) = 14.5 + 14.5 + 1 + (0.18 * 14.5 * 8.5) + 1.6 = 53.785
Which should overheat the mech

According to Li Song, the total heat generated by this build after an alpha strike is:
Total heat (Li Song) = 14.5 + 14.5 + 1 + (0.18 * 14.5 * 7) + 1.6 = 49.87
which should not create an overheat.

The following recording shows that an alpha strike is crossing the 50 heat threshold causing an overheat, indicating that the total heat is higher than 50.
https://youtu.be/WvEVDDijREg


To confirm the value calculated with the correct multiplier (53.785 total alpha heat), the same Vapor eagle used above is equipped with 8 additional double heatsinks to bring up the total amount of heat capacity to 54, reaching just above the alpha strike heat. In this case an alpha strike should not cause an overheat.
VGL-1 with added heatsinks: https://mwo.nav-alpha.com/mechlab?b=f70662ff_VGL-1

The following recording shows that the VGL with the additional heatsinks is able to alpha strike without an overheat:
https://youtu.be/7T3OeZ2z4M0

**Comment by NavidA1** [comment]
It should also be noted that moving no longer subtracts from heat capacity as part of the change made to heat systems in 2017. Moving only generates additional heat per second (around 0.3 heat per second).

Same test with the Madcat MKII above... but moving this time:
https://youtu.be/OJUtkMml0yw

**Comment by LiSongMWO** [comment]
Related question: 

Suppose you have a mech that has a "heat scale limit" modifier to a weapon like for example AWS-8Q that allows you to fire 3 PPCs without ghost heat. Let's for a moment assume that the mech could equip a gauss rifle, which shares group with PPCs; Would this allow you to fire 2 PPC and 1 gauss without ghost heat? Or is it only 3 PPC that are exempt? Because that calculation becomes a bit cumbersome.

**Comment by LiSongMWO** [comment]
It seems that VTR-DS is just such a mech with a `PPC FAMILY HSL +1` quirk, which despite the wording `PPC FAMILY` also applies to the gauss rifle, allowing you to fire two PPCs and a Gauss without ghost heat, if the ingame mechlab is to be believed.

Based on some sciencing in the ingame mechlab, it would appear as if for each weapon in the group the maximum free alpha is calculated individually and applying any quirks individually. The whole group then considers the weapon with the lowest free alpha. So even though you could technically fire 3 light PPCs without ghost heat as they have max free alpha set to 3, you can't fire 2 light PPCs and one ER PPC because ER PPC only has 2 max free alpha....

**Comment by NavidA1** [comment]
That's correct. It is how the het penalty system works.

It searches for the weapon with the lowest allowable limit in the group and if the total number of the weapons in the group exceeds that number, it applies penalty, using the normal rules of heat penalty system (i.e., highest heat of the group, along with highest penalty multiplier of the group)

Gauss is in the same group as PPCs. In a change last year, the heat penalty limit of gauss was changed to 7 (way above normal limits of PPCs), to put the penalty gate on the PPCs themselves, and allow the game to control how many Gauss / PPC combos you can fire just using PPC family HSL, without having to add Gauss HSL at the same time as well.
Gauss limit at 7, or 999 by itself works because you can't charge more than 2 Gauss rifles regardless.
